cluster-up: Unify detection of podman or docker as runtime

0xFelix opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

These changes unify the detection of podman or docker as runtime. This
makes 'make cluster-up' and 'make cluster-sync' always use the same
runtime. Prior to these changes 'make cluster-up' would have used docker
while 'make cluster-sync' would have used podman if both were installed
at the same time and podman.socket was not available.

These changes also set the path of the podman's socket to
'${XDG_RUNTIME_DIR}/podman/podman.socket' so podman can be started with
'systemctl start --user podman.socket' which is the standard way of
starting podman as user on distributions using systemd.

**Which issue(s) this PR fixes**

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

0xFelix commented:
See also https://github.com/kubevirt/kubevirtci/pull/741

0xFelix commented on hack/common.sh:
```diff
@@ -6,7 +6,7 @@ determine_cri_bin() {
         echo docker
     else
-        if podman ps >/dev/null 2>&1; then
+        if curl --unix-socket "${XDG_RUNTIME_DIR}/podman/podman.sock" http://d/v3.0.0/libpod/info >/dev/null 2>&1; then
```
Checking for the socket here makes the logic for determining whether to use `podman` or `docker` uniform across the `make` targets. Without the PR `make cluster-up` would use `docker` and `make cluster-sync` would use `podman`, if both are installed at the same time and `podman.sock` is not available.

In it's default configuration the `podman.sock` is not available as `podman` needs no daemon by default. It must be explicitly enabled. The PR makes `podman` opt-in from the user's perspective. `docker` always runs a daemon and therefore `docker ps` will not work without an accessible `docker` socket.

I thought of using `podman.sock` as a convenient way to switch between rootless and rootful `podman`. However I just noticed there are instances where `KUBEVIRT_CRI` is called directly. Calling `KUBEVIRT_CRI` directly means always using rootless `podman` unless `make` is called as root. This could be alleviated by passing the `podman.sock` to `podman` (as `--url` argument). What do you think?

0xFelix commented:
@iholder-redhat I tested it locally on FC35 and it worked fine with rootful podman. Rootless podman still had networking problems during `make cluster-up`. Using the `podman.sock` is a convenient way to switch between rootless and rootful `podman`. [1]

[1] https://github.com/kubevirt/kubevirtci/blob/main/PODMAN.md

0xFelix commented:
/retest

orelmisan commented:
Hi @0xFelix, thanks for the PR.
What are the preliminary steps required to run `make cluster-up` and `make cluster-sync` using podman?
I've found [this](https://github.com/kubevirt/kubevirtci/blob/main/PODMAN.md) document you co-wrote.

0xFelix commented:
Hi @orelmisan 

Right, the document outlines all necessary steps. It could be included in the `kubevirt/kubevirt` repo too?

orelmisan commented:
IMO it would be nice to have it in [getting-started](https://github.com/kubevirt/kubevirt/blob/main/docs/getting-started.md) or some other file under the docs directory.

@enp0s3 could you please provide guidance?

EdDev commented:
/retest

EdDev reviewed (approved):
Thank you!

0xFelix commented:
I changed to PR so every call to podman uses a socket. This guarantees that rootless and rootful podman are not mixed. It is up to the user to provide a podman.socket which is either rootless or rootful. See also https://github.com/kubevirt/kubevirtci/pull/757.

Additionally I added a note to [getting-started](https://github.com/kubevirt/kubevirt/blob/main/docs/getting-started.md).

enp0s3 commented:
/assign

EdDev commented on hack/dockerized:
```diff
@@ -148,8 +147,8 @@ fi
 test -t 1 && USE_TTY="-it"
-EXEC_ENV_VARS="--env KUBEVIRT_CRI=${KUBEVIRT_CRI}"
-if ! $KUBEVIRT_CRI exec $EXEC_ENV_VARS ${USE_TTY} ${BUILDER}-bazel-server /entrypoint.sh "$@"; then
+EXEC_ENV_VARS=("--env" "KUBEVIRT_CRI=${KUBEVIRT_CRI}")
```
Why is this change needed?

0xFelix replied:
It is needed because of spaces introduced by using `podman` with the `--remote` and `--url` arguments here https://github.com/kubevirt/kubevirt/blob/9c31510e3f22991c7cfcc232ef67a4b6d16134b1/hack/common.sh#L10

Although I am sure why the `$KUBEVIRT_CRI` as env variable is needed here in the first place. Can you explain? @iholder-redhat

iholder-redhat replied:
I think the purpose of `KUBEVIRT_CRI` is to explicitly specify a CRI.
IOW, if you have both Podman and Docker installed and you want to explicitly state that you want to use one of them.

0xFelix replied:
Yes, but why does the `entrypoint.sh` of bazel-server need `KUBEVIRT_CRI`?

enp0s3 replied:
@0xFelix Hi. I think its needed for the other bash script file provided from the outer context and that is invoked by `"$@"` after `entrypoint.sh` exeution.

0xFelix replied:
Hi @enp0s3, sorry, I see! `entrypoint.sh` runs a build script. Are podman and docker included in the builder image? Sorry, I am confused.

enp0s3 replied:
@0xFelix Lets remove it, but please run `make` and `make test` to make sure we don't break anything. Thanks!

0xFelix replied:
@enp0s3 `make` and `make test` run sucessfully after removing `EXEC_ENV_VARS`. I updated the PR.

EdDev reviewed (approved):
Thank you!

enp0s3 commented:
/approve

0xFelix commented:
/retest

oshoval commented:
Great work thanks @0xFelix 
Would you have the honors please to update https://github.com/kubevirt/kubevirtci/blame/main/PODMAN.md#L11 ?
It says it doesn't work with rootless still but it does thanks to this PR

0xFelix commented:
Thank you @oshoval :+1: 

Unfortunately the documentation is right. It does not work 100% with rootless podman. :(
The cluster is not able to access external resources when running in rootless podman. However as long as the cluster does not need external resources it works fine.
